projects/healthPingValidation.kp/knowledge.md

```diff
@@ -0,0 +1,550 @@
+
+def getAvgPerDate(iterable):
+    aggregare = Counter([d for d, r in iterable])
+    result = sum(aggregare.values()) * 1.0 / len(aggregare)
+    return result
+
+def pingsPerClientPerDay(pings, date):
+    return pings.map(lambda ping: (ping["clientId"], (ping[date], ping["payload/reason"]))).groupByKey()
+
+def avgPingsPerClientPerDay(pings, date):
+    idDateRDD = pingsPerClientPerDay(pings, date)
+    return idDateRDD.map(lambda pair: (pair[0], getAvgPerDate(pair[1]))).collect()
+```
+
+```python
+def plotAvgPingPerDateDistr(pings, date):
+    PINGS_COUNT_PER_DAY = 24
+    resultDistributionList = avgPingsPerClientPerDay(pings, date)
+    values = [v for k, v in resultDistributionList]
+    plt.title(date + ": clients sending more than expected ping per day - " + \
+                  str(len([v for v in values if v > PINGS_COUNT_PER_DAY])) + \
+                      ". Log on Y axis")
+    plt.yscale('log')
+    plt.plot(sorted(values))  
+    plt.show()
+```
+
+```python
+plotAvgPingPerDateDistr(cachedData, "meta/submissionDate")
+plotAvgPingPerDateDistr(cachedData, "creationDate")
+```
+
+
+![png](images/output_31_0.png)
+
+
+
+
+![png](images/output_31_1.png)
+
+
+Turns out, clients submit health pings properly (less that 24/day) but we get them on server with some delay
+
+# Nightly 56 & 57. Daily active Health ping clients against DAU
+
+DAU57, DAU56 from re:dash https://sql.telemetry.mozilla.org/queries/15337/source#table
+
+
+```python
+DAU57 = [('20170802', 9497), ('20170803', 20923), ('20170804', 25515), ('20170805', 24669), ('20170806', 25604), \
+         ('20170807', 32762), ('20170808', 36011), ('20170809', 37101), ('20170810', 38934), ('20170811', 38128), \
+         ('20170812', 33403), ('20170813', 33301), ('20170814', 38519), ('20170815', 39053), ('20170816', 41439), \
+         ('20170817', 41482), ('20170818', 40557), ('20170819', 34310)]
+DAU56 = [('20170727', 34419), ('20170728', 33142), ('20170729', 27740), ('20170730', 28194), ('20170731', 34340), ('20170801', 37308)]
+```
+
+```python
+def getClientsPerDay(pings, version, dateFrom, dateTo):
+    clientsPerDay = pings.filter(lambda ping: ping["application/version"] == version) \
+        .map(lambda ping: (ping["meta/submissionDate"], ping["clientId"])) \
+        .groupByKey() \
+        .map(lambda pair: (pair[0], len(set(pair[1])))).collect()
+        
+    return filter(lambda pair: pair[0] >= dateFrom and pair[0] <= dateTo, clientsPerDay)
+
+filtered57 = getClientsPerDay(cachedData, '57.0a1', '20170801', '20170819')
+filtered56 = getClientsPerDay(cachedData, '56.0a1', '20170727', '20170801')
+```
+
+```python
+plotlistofTuples(sorted(DAU56 + DAU57), inColor='red')
+plotlistofTuples(sorted(filtered56 + filtered57), "Nightly 56&57. 56 from 20170727 to 20170801. 57 from 20170802 to 20170815")
+plt.legend(['DAU', 'Health ping clients/day'])
+
+```
+
+
+
+
+    <matplotlib.legend.Legend at 0x7f0650f5dfd0>
+
+
+
+
+
+![png](images/output_37_1.png)
+
+
+
+```python
```

Comment:
It would be nice to have a conclusion from this DAU/health ping client stacked plot. Even if it is a wild guess about what it means (as long as you mention it is "speculation") it will encourage others to think about it.